### PR TITLE
Db 843 file f routes

### PR DESCRIPTION
**Pull request by jworcestBAH**
Created consolidated routes and functions for files D1, D2, E, and F.  This PR does not remove the existing routes, so it does not require coordination with frontend deploy.  A later PR will remove the D-specific functionality.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -38,3 +38,4 @@ shared-local-instance.db
 dataactcore/scripts/dropMany.py
 runDynamo.py
 dataactvalidator/config/sf_133.csv
+runTests.py
```
Comment (cmc333333) [inline .gitignore]: We don't want to check in changes to runTests.py?

File: dataactbroker/handlers/fileHandler.py [hunk 2]
```diff
@@ -163,14 +164,17 @@ def submit(self,name,CreateCredentials):
...
+                # Don't add external files to existing submission
                 for extFileType in FileHandler.EXTERNAL_FILE_TYPES:
                     if extFileType == "award":
                         filename = CONFIG_BROKER["d2_file_name"]
                     elif extFileType == "award_procurement":
                         filename = CONFIG_BROKER["d1_file_name"]
+                    elif extFileType == "awardee_attributes":
+                        filename = CONFIG_BROKER["awardee_file_name"]
+                    elif extFileType == "sub_award":
+                        filename = CONFIG_BROKER["sub_award_file_name"]
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Not a big deal, but now that there are four branches, would it make sense to move this into a mapping? Something like:
```python
FILE_TYPE_TO_NAME = {
    "award": CONFIG_BROKER["d2_file_name"],
    "award_procurement": CONFIG_BROKER["d1_file_name"],
    ...
}
...
filename = FILE_TYPE_TO_NAME[extFileType]
```

It'd also be a good idea to standardize these names (if possible) so that we could use, e.g.
```python
filename = CONFIG_BROKER[extFileType + "_file_name"]
```

File: dataactbroker/handlers/fileHandler.py [hunk 3]
```diff
@@ -297,8 +301,16 @@ def getStatus(self):
             for jobId in jobs:
                 jobInfo = {}
                 jobType = self.jobManager.getJobType(jobId)
+
                 if jobType != "csv_record_validation" and jobType != "validation":
                     continue
+
+                # TODO Skip D1 file until validation is added, remove these lines once D1 validation is added
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Is there a way to make this more prominent? Maybe make a separate issue if one doesn't already exist? I don't think anyone's going to see this comment if they aren't specifically looking for it.

File: dataactbroker/handlers/fileHandler.py [hunk 4]
```diff
@@ -406,19 +418,22 @@ def uploadFile(self):
...
-            return JsonResponse.create(exc, exc.status)
+            return JsonResponse.error(exc, exc.status)
 
         submission_id = requestDict.getValue("submission_id")
         start_date = requestDict.getValue("start")
         end_date = requestDict.getValue("end")
-
+        job = jobDb.getJobBySubmissionFileTypeAndJobType(submission_id, self.EXTERNAL_FILE_TYPE_MAP["D1"], "file_upload")
+        job.start_date = datetime.strptime(start_date,"%m/%d/%Y").date()
+        job.end_date = datetime.strptime(end_date,"%m/%d/%Y").date()
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Do we need any validation here? If the data comes from the user, do we risk and explosion?

Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Unrelated, but it looks like this job isn't used until lines 448. Would it make sense to move this creation logic closer to where the job's used?

File: .gitignore (hunk 1 above)
Comment (jworcestBAH) [inline .gitignore]: That's a different runTests, I was trying to have a script that would send the pytest shell command to run both the integration and unit tests, and I assumed we didn't want that in the repo.  I'll rename this to something less confusing.

File: dataactbroker/handlers/fileHandler.py (hunk 3 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: I will add my notes on the D1 changes that are needed to DB-925, I have a list of the places that got these TODO notes.

File: dataactbroker/handlers/fileHandler.py (hunk 4 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: The frontend has users pick from a calendar that sends the dates in that format, so there is no free entry.  If a bad format is sent, this will raise an exception and return a 500.  We could catch that and return a 400 with a more generic message if that would be preferable.

Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: These lines are storing the data in the jobs table to be returned in the later call to /check_generation_status/.  448 stores other info in the jobs table for the same purpose.  My thinking in putting these lines here was to be closer to where start_date and end_date are created so it's clearer what is being stored.  That said, I can see how it could be clearer to have all the job-related lines in one place, so I'll move these down to 448.

File: dataactbroker/handlers/fileHandler.py [hunk 5]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
 
         return JsonResponse.create(StatusCode.OK, response)
 
+    def loadGenerateRequest(self):
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Not a big deal, but I found this method name confusing. Would something like `getSubmissionAndFileType` be too on the nose?

File: dataactbroker/handlers/fileHandler.py [hunk 6]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
...
+                                    StatusCode.CLIENT_ERROR)
+            return JsonResponse.error(exc, exc.status)
+
+        submission_id = requestDict.getValue("submission_id")
+        file_type = requestDict.getValue("file_type")
+        return submission_id, file_type
+
+    def generateFile(self):
+        """ Start a file generation job for the specified file type """
+        submission_id, file_type = self.loadGenerateRequest()
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Will this explode if the request is not formatted? `loadGenerateRequest` will return a single element, not a tuple.

File: dataactbroker/handlers/fileHandler.py [hunk 7]
```diff
@@ -479,9 +497,134 @@ def checkD1File(self, submission_id=None):
...
+                        result = "failed"
+                        uploadJob.error_message = "Validation completed but row-level errors were found"
+                    else:
+                        result = "finished"
+        self.interfaces.jobDb.session.commit()
+        return result
+
 
     def generateD2File(self):
         """ Initiates the generation of D2 """
```
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Incidentally, there is a good bit of repeated code here from the D1 function.  The next PR will be consolidating these along with E and F to reduce that replication.  The reason that's being done as a separate PR is that it will require more coordination with the frontend to ensure the routes being removed are no longer used.

File: dataactbroker/handlers/fileHandler.py [hunk 8]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
...
+        """ Start a file generation job for the specified file type """
+        submission_id, file_type = self.loadGenerateRequest()
+        job = self.interfaces.jobDb.getJobBySubmissionFileTypeAndJobType(submission_id, self.EXTERNAL_FILE_TYPE_MAP[file_type], "file_upload")
+        # Check prerequisites on upload job
+        if not self.interfaces.jobDb.runChecks(job.job_id):
+            exc = ResponseException("Must wait for completion of prerequisite validation job", StatusCode.CLIENT_ERROR)
+            return JsonResponse.error(exc, exc.status)
+        if file_type == "D1":
+            # TODO could this function return S3 file location
+            result = self.generateD1File()
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Newbie question, but it looks like all of the communications between these functions is implicit via the shared `self.request` object. Given that we just derived the `job` and `submission_id`, though, would it make sense to pass them along? That'd make writing unit tests around `generateD1File` significantly simpler, too.

File: dataactbroker/handlers/fileHandler.py [hunk 9]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
...
+            exc = ResponseException("File type must be either D1, D2, E or F", StatusCode.CLIENT_ERROR)
+            return JsonResponse.error(exc, exc.status)
+
+        # TODO mark S3 file location in DB
+        # Mark file generation upload as finished
+        self.interfaces.jobDb.markJobStatus(job.job_id,"finished")
+        # Return same response as check generation route
+        return self.checkGeneration()
+
+    def checkGeneration(self):
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Similar here: why not pass in the `submission_id` and `file_type` rather than re-deriving it?

File: dataactbroker/handlers/fileHandler.py [hunk 10]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
...
+        responseDict["status"] = self.mapGenerateStatus(uploadJob, validationJob)
+        responseDict["file_type"] = file_type
+        responseDict["message"] = uploadJob.error_message or ""
+        if uploadJob.filename is None:
+            responseDict["url"] = ""
+        else:
+            if CONFIG_BROKER["use_aws"]:
+                responseDict["url"] = s3UrlHandler().getSignedUrl("d-files",uploadJob.filename, bucketRoute=None, method="GET")
+            else:
+                responseDict["url"] = uploadJob.filename
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: No biggie, but these nested ifs can be collapsed to be less pointy:
```python
if uploadJob.filename is None:
    ...
elif CONFIG_BROKER["use_aws"]:
   ...
else:
   ...
```

File: dataactbroker/handlers/fileHandler.py (hunk 6 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Good catch, changed loadGenerateRequest to raise an exception in that case.  Exceptions get translated to the appropriate response object in dataactbroker.routeUtils

File: dataactbroker/handlers/fileHandler.py [hunk 11]
```diff
@@ -479,9 +498,134 @@ def checkD1File(self, submission_id=None):
...
+                # If validation job exists, check that status as well
+                if validationStatus in ["waiting", "ready", "running"]:
+                    result = "waiting"
+                elif validationStatus in ["invalid", "failed"]:
+                    if validationJob.error_message is None:
+                        if uploadJob.error_message is None:
+                            if validationStatus == "invalid":
+                                uploadJob.error_message = "Generated file had file-level errors"
+                            else:
+                                uploadJob.error_message = "Validation job had an internal error"
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: The logic here is nested _very_ deeply (we're at 6 levels of `if`). Is there a way to simplify the logic so that it's easier to follow? Right now I need to effectively execute the function in my head to figure out which path is taken.

File: dataactbroker/handlers/fileHandler.py (hunk 5 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Only thing there is that it looks a lot like our naming of interface functions to pull from DB.  I was hoping to make this clear that it's pulling from the http request, would something like getRequestParams make more sense, maybe getRequestParamsForGenerate since this only applies to the file generation routes?

File: dataactbroker/handlers/fileHandler.py (hunk 4 above)
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Trusting JavaScript for input validation is asking for trouble; is it safe to say the rest of the application does this? We can make a separate issue if so.

File: dataactbroker/handlers/fileHandler.py (hunk 8 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Right, this is in an awkward halfway state where generateD1File and generateD2File can be called directly from the old routes.  The next PR will be removing those routes and consolidating the generateD1File and generateD2File functions into a more generic function that handles E and F as well.  The consolidated function will certainly take submission ID and file type as arguments, probably the job as well depending on what it returns.

File: dataactbroker/handlers/fileHandler.py (hunk 9 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: I could make those optional arguments here, but as this function is the first called from the /check_generate_status/ route it needs to be able to pull the values from the request if they're not provided.

File: dataactcore/aws/s3UrlHandler.py [hunk 12]
```diff
@@ -43,6 +43,7 @@ def _signUrl(self,path,fileName,bucketRoute,method="PUT") :
         """
         if(s3UrlHandler.ENABLE_S3) :
             s3connection = boto.s3.connect_to_region(s3UrlHandler.REGION)
+            print(str(s3connection))
```
Comment (cmc333333) [inline dataactcore/aws/s3UrlHandler.py]: Looks like a stray debug line

File: dataactbroker/handlers/fileHandler.py (hunk 9 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Added that option.

File: dataactbroker/handlers/jobHandler.py [hunk 13]
```diff
@@ -259,6 +259,8 @@ def addUploadJobs(self,filenames,submissionId,existingSubmission):
         jobsRequired = []
         # Dictionary of upload ids by filename to return to client
         uploadDict = {}
+        d1ValId = None
+        cValId = None
```
Comment (cmc333333) [inline dataactbroker/handlers/jobHandler.py]: If I understand the logic correctly, these vars won't be set unless the `filenames` are in a specific order. Is there a way to be more explicit? Maybe we pull the file C and D1 files out of that list and run them first?

File: dataactbroker/handlers/fileHandler.py (hunk 10 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Thanks, fixed

File: dataactcore/models/jobTrackerInterface.py [hunk 14]
```diff
@@ -229,6 +230,62 @@ def checkJobDependencies(self,jobId):
...
+        Args:
+        jobId -- ID of job to be checked
+        Returns:
+        True if all passed, otherwise False or exception
+        """
+        if(self.checkJobReady(jobId) and self.checkPrerequisites(jobId)):
+            # All passed
+            return True
+        else:
+            return False
```
Comment (cmc333333) [inline dataactcore/models/jobTrackerInterface.py]: Might be useful to simplify to:
```python
return self.checkJobReady(jobId) and self.checkPrerequisites(jobId)
```

File: dataactcore/aws/s3UrlHandler.py (hunk 12 above)
Comment (jworcestBAH) [inline dataactcore/aws/s3UrlHandler.py]: Yep, removed

File: dataactcore/models/jobTrackerInterface.py [hunk 15]
```diff
@@ -229,6 +230,62 @@ def checkJobDependencies(self,jobId):
...
+        result = self.checkJobUnique(query)
+        # Found a unique job
+        if(result.job_status_id != self.getJobStatusId("ready")):
+            # Job is not ready
+            # Job manager is not yet implemented, so for now doesn't have to be ready
+            return True
+            # TODO when job manager exists, change to exception below
+            #exc = ResponseException("Job is not ready",StatusCode.CLIENT_ERROR,None,ValidationError.jobError)
+            #exc.status = 400
+            #raise exc
```
Comment (cmc333333) [inline dataactcore/models/jobTrackerInterface.py]: Not a big deal, but rather than commenting all of this code out, it might be cleaner to just exit early, e.g.
```python
def checkJobReady(self, jobId):
    # Comment explaining why we're exiting early
    return True
    query = ...
```
I think the benefit there is that readers don't need to look through branches to figure out that the function is only stubbed.

File: dataactcore/models/jobTrackerInterface.py [hunk 16]
```diff
@@ -229,6 +230,62 @@ def checkJobDependencies(self,jobId):
...
+
+        return True
+
+    def checkPrerequisites(self, jobId):
+        """ Checks that specified job has no unsatisfied prerequisites
+        Args:
+        jobId -- job_id of job to be run
+
+        Returns:
+        True if prerequisites are satisfied, False otherwise
```
Comment (cmc333333) [inline dataactcore/models/jobTrackerInterface.py]: Note that this doesn't actually return False if the prereqs aren't satisfied; it raises an exception

File: dataactcore/models/jobTrackerInterface.py (hunk 15 above)
Comment (jworcestBAH) [inline dataactcore/models/jobTrackerInterface.py]: So this is much older code that is just being moved into the core in this PR.  Actually the job manager is now implemented, but it has been useful in the past to be able to rerun validations on a finished job, so I think I will remove this check entirely.

File: dataactcore/models/jobTrackerInterface.py [hunk 17]
```diff
@@ -229,6 +230,62 @@ def checkJobDependencies(self,jobId):
...
+        """
+        # Get list of prerequisites
+        queryResult = self.session.query(JobDependency).filter(JobDependency.job_id == jobId).all()
+        for prereq in queryResult:
+            query = self.session.query(Job).filter(Job.job_id == prereq.prerequisite_id)
```
Comment (cmc333333) [inline dataactcore/models/jobTrackerInterface.py]: It looks like our ORM models aren't using Sqlalchemy's [relationships](http://docs.sqlalchemy.org/en/latest/orm/tutorial.html#building-a-relationship), which would make this a bit simpler. Rather than crafting our own queries, relationships push that logic into the ORM.

Not worth holding anything back, but worth considering.

Comment (cmc333333) [inline dataactcore/models/jobTrackerInterface.py]: Somewhat related, would it be worthwhile to just move all of this Python logic into the query? I don't know the Sqlalchemy functions well enough, but using a made-up syntax:
```python
query = self.session.query(JobDependency).join(Job, on=JobDependency.job_id == Job.job_id).filter(
    Job.status_id=="finished")
if query.exists():
    raise Exception
```

File: tests/integration/fileTests.py [hunk 18]
```diff
@@ -433,6 +438,61 @@ def test_metrics(self):
...
+        self.assertEqual(json["status"], "failed")
+        self.assertEqual(json["file_type"], "F")
+        self.assertEqual(json["url"],"")
+        self.assertEqual(json["message"],"File was invalid")
+
+        # E file generation should error because D1 job did not validate
+        postJson = {"submission_id": self.generation_submission_id, "file_type": "E", "start":"01/02/2016", "end":"02/03/2016"}
+        response = self.app.post_json("/v1/generate_file/", postJson, headers={"x-session-id":self.session_id}, expect_errors=True)
+
+        print("Response: " + str(response.json))
```
Comment (cmc333333) [inline tests/integration/fileTests.py]: Looks like a stray debug line

File: dataactcore/models/jobTrackerInterface.py (hunk 16 above)
Comment (jworcestBAH) [inline dataactcore/models/jobTrackerInterface.py]: Fixed the docstring

File: dataactbroker/fileRoutes.py [hunk 19]
```diff
@@ -102,3 +102,17 @@ def get_protected_files():
...
+
+    @app.route("/v1/generate_file/", methods=["POST"])
+    @permissions_check
+    def generate_file():
+        """ Generate file from external API """
+        fileManager = FileHandler(request, isLocal=IS_LOCAL, serverPath=SERVER_PATH)
+        return RouteUtils.run_instance_function(fileManager, fileManager.generateFile)
+
+    @app.route("/v1/check_generation_status/", methods=["POST"])
+    @permissions_check
```
Comment (cmc333333) [inline dataactbroker/fileRoutes.py]: Maybe out of scope, but I would assume that a "check" function would respond to a `GET` rather than a `POST`. I'd expect something like:

http://example.com/v1/generation_status/D1?session=mytoken

File: dataactbroker/handlers/fileHandler.py (hunk 5 above)
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Both options sound better to me

**Comment by cmc333333** [comment]
Sorry for the delay; done reviewing now. I added lots of comments, but they're all take-em-or-leave-em. The one thing we do need to get sorted out before merging are the test failures

File: dataactbroker/handlers/fileHandler.py (hunk 2 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Changed to use the second approach

File: dataactbroker/handlers/fileHandler.py (hunk 4 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Changed to return a 400 and raise a ResponseException rather than a 500.

**Comment by jworcestBAH** [comment]
The test failures are from Jenkins having an older version of the config file, will update that after I get through the other changes.

File: dataactbroker/handlers/fileHandler.py (hunk 5 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Okay, changed

File: dataactbroker/handlers/fileHandler.py (hunk 11 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/fileHandler.py]: Simplified this section using a couple dictionary lookups, should be clearer now.

File: dataactcore/models/jobTrackerInterface.py (hunk 14 above)
Comment (jworcestBAH) [inline dataactcore/models/jobTrackerInterface.py]: This got simplified down to a single function with no ready check

File: dataactcore/models/jobTrackerInterface.py (hunk 17 above)
Comment (jworcestBAH) [inline dataactcore/models/jobTrackerInterface.py]: For relationships, we are using those in some places, but they are only populated if you explicitly do a joinedload or a subqueryload, so I have mostly been avoiding relying on them as they will be empty when passing around an instance created without those.  That's not the case here, so we could use this here, I'll take a look after I figure out the C and D1 issue above.

File: tests/integration/fileTests.py (hunk 18 above)
Comment (jworcestBAH) [inline tests/integration/fileTests.py]: Removed

File: dataactbroker/fileRoutes.py (hunk 19 above)
Comment (jworcestBAH) [inline dataactbroker/fileRoutes.py]: These routes were defined in https://gist.github.com/kevinli-work/59f7310afc789fdeeb7b0993772df03d, so I followed what was in that.  This route does take a couple parameters, which is why it's a post route.  We could make this a GET route and have the parameters be part of the URL, but that isn't really consistent with how we've done similar routes like /v1/check_status/.  If we want to change this pattern, it should probably be its own story to go through and identify the routes that should be done this way, with hours included for the necessary frontend changes to use the altered routes.

File: dataactbroker/handlers/jobHandler.py (hunk 13 above)
Comment (jworcestBAH) [inline dataactbroker/handlers/jobHandler.py]: Refactored this to run C and D1 before the rest

File: dataactcore/models/jobTrackerInterface.py (hunk 17 above)
Comment (jworcestBAH) [inline dataactcore/models/jobTrackerInterface.py]: Refactored to use relationships

**Comment by jworcestBAH** [comment]
Finished with my changes, let me know if there's anything I missed.

File: dataactbroker/handlers/fileHandler.py [hunk 20]
```diff
@@ -576,47 +578,42 @@ def mapGenerateStatus(self, uploadJob, validationJob = None):
...
-                elif validationStatus == "finished":
-                    if errorsPresent:
-                        result = "failed"
-                        uploadJob.error_message = "Validation completed but row-level errors were found"
-                    else:
-                        result = "finished"
+                uploadJob.error_message = "Validation job had an internal error"
+
+        elif uploadJob.error_message is None:
+            uploadJob.error_message = validationJob.error_message
```
Comment (cmc333333) [inline dataactbroker/handlers/fileHandler.py]: Very nice refactor! The branching is easier to follow now.

**Comment by cmc333333** [comment]
Lots of back and forth here; thanks for tolerating my digressions! Better thanks for writing tests :)
